src/internal_modules/roc_sndio/pump.cpp

```diff
@@ -81,6 +83,26 @@ void Pump::stop() {
     stop_ = 1;
 }
 
+status::StatusCode Pump::close_device_(IDevice* device) {
+    if (device) {
+        return device->close();
+    }
+    return status::StatusOK;
+}
+
+status::StatusCode Pump::close_all_devices_() {
+    IDevice* devices[] = { &main_source_, &sink_, backup_source_, current_source_ };
+
+    for (size_t i = 0; i < sizeof(devices) / sizeof(devices[0]); ++i) {
+        status::StatusCode device_code = close_device_(devices[i]);
+        if (device_code != status::StatusOK) {
+            return device_code;
+        }
```

Comment:
If close failed for one of the devices, we should still close remaining devices, instead of early return. Ideally we should return the very first error (if there were errors).

Also please log close failure with status (see status::code_to_str).